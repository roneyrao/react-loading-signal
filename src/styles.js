export default from './styles/FlexLoading.css';
