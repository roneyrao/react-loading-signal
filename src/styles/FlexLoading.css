.loading {
	width:100%;
	top:50%;
	transform: translateY(-50%);
	z-index:999;
	visibility:hidden;
	text-align:center;
	word-wrap:normal;
	white-space:nowrap;
}
.masked{
	top:0;
	right:0;
	bottom:0;
	left:0;
	z-index:9999;
	display:flex;
	justify-content:center;
	align-items:center;
	transform: none;
}
.masked:before{
	content:'';
	top:0;
	right:0;
	bottom:0;
	left:0;
	background:rgba(150, 150, 150, 0.4);
	position: absolute;
}
.masked.loading>* {
	transform: none;
	z-index: 1;
}

.local{
	composes:loading;
	position:absolute;
}
.localMasked{
	composes:local masked;
}

.global{
	composes:loading;
	position:fixed;
}
.globalMasked{
	composes:global masked;
}

.messageList{
	margin-top:1em;
	line-height:2em;
}
/*theme 'blobs'*/
.blobs{
    width: 100%;
	height:1em;
	background:url(./images/blobs.gif) no-repeat center center;
	background-size:auto 100%;
}
/*theme 'spinning'*/
.spinning{
	min-width:3em;
    padding: 4.5em 1.5em 1.5em;
    border-radius: 0.8em;
	background:rgba(0,0,0,.7) url(./images/spinning.gif) no-repeat center 1.5em;
	background-size:auto 3em;
	color: white;
}





