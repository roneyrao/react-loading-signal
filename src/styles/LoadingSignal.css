.loading {
  top: 50%;
  left: 50%;
  z-index: 999;
  visibility: hidden;
  height: 0;
  text-align:center;
}
.loading > * {
  display: inline-block;
  transform: translate(-50%, -50%);
}
.masked {
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 9999;
  display: flex;
  justify-content: center;
  align-items: center;
  transform: none;
}
.masked:before {
  content: '';
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background: rgba(150, 150, 150, 0.4);
  position: absolute;
}
.masked.loading > * {
  transform: none;
  z-index: 1;
}

.local {
  composes: loading;
  position: absolute;
}
.localMasked {
  composes: local masked;
}

.global {
  composes: loading;
  position: fixed;
}
.globalMasked {
  composes: global masked;
}

.messageList {
  line-height: 1em;
  list-style: none;
  margin:0;
  padding:0;
}
.messageList > li {
  padding-top: 1em;
}
/*theme 'blobs'*/
.blobs {
  width: 12em;
  height: 1em;
  background: url(./images/blobs.gif) no-repeat center center;
  background-size: auto 100%;
}
/*theme 'spinning'*/
.spinning {
  min-width: 3em;
  max-width: 50%;
  padding: 4.5em 1.5em 1.5em;
  border-radius: 0.8em;
  background: rgba(0, 0, 0, 0.7) url(./images/spinning.gif) no-repeat center 1.5em;
  background-size: auto 3em;
  color: white;
  text-align: center;
  word-wrap: break-word;
}
/*theme 'progress'*/
.progress {
}
.progressBar {
  height: 5px;
  border: 1px solid #057927;
}
.progressBar > div{
  width: 0%;
  height: 100%;
  background: rgb(5, 121, 39);
}
