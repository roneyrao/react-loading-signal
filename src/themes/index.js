export Blobs from './blobs';
export Spinning from './spinning';
export Progress from './progress';
