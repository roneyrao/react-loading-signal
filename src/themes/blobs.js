import React from 'react';
import styles from '../styles';
export default function LoadingBlobs(){
	return (<div className={styles.blobs}/>);
}
