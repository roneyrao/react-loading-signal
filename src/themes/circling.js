// @flow
import React from 'react';
import styles from '../styles';

export default function LoadingSpinning({ message }: { message?: React.ChildrenArray<mixed> }) {
  return <div className={styles.circling}>{message}</div>;
}
LoadingSpinning.defaultProps = {
  message: 'loading...',
};
