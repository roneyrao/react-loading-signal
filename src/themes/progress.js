// @flow
import React from 'react';
import styles from '../styles';

export type Caption = (number) => string;
export type ProgressProps = {
  message: number,
  caption: Caption,
};


function defaultCaption(progress) {
  return `progress: ${(progress * 100).toPrecision(4)}%`;
}

export default function LoadingProgress({ message, caption }: ProgressProps) {
  return (
    <div className={styles.progress}>
      <div>{(caption || defaultCaption)(message)}</div>
      <div className={styles.progressBar}>
        <div style={{ width: `${message * 100}%` }} />
      </div>
    </div>
  );
}
