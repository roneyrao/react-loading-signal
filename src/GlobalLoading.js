/**
 * Created by roney on 2017/9/20.
 */

import React from 'react';
import { render } from 'react-dom';
import PropTypes from 'prop-types';
import { Spinning } from './themes';
import { LoadingEvent, EVENT_LOCAL_HADNDLED } from './LoadingEvent';
import styles from './styles';

// can theme be changed? can multiple themes coexist? no! global is singleton, one sign for many loadings.
export class GlobalLoadingComp extends React.Component {
  static propTypes = {
    theme: PropTypes.func,
    masked: PropTypes.bool,
    closable: PropTypes.bool,
  };
  static defaultProps = {
    theme: Spinning,
    masked: false,
    closable: false, // if the masked overlay can be closed when click;
  };
  state = {
    shown: false,
    messages: {},
  };

  loadingCount = 0; // multiple loadings may be pending meantime;

  generateID() {
    return `loading${new Date().valueOf()}${Math.round(Math.random() * 10000)}`;
  }
  addLoading(msg = '') {
    this.loadingCount++;
    const id = this.generateID();
    const newState = { shown: true };
    if (msg) {
      newState.messages = { ...this.state.messages, [id]: msg };
    }
    this.setState(newState);
    return id;
  }
  removeLoading(id) {
    // console.log('remove loading', id);
    this.loadingCount--;
    if (this.loadingCount < 0) {
      // unknown exception;
      // console.error('loading removed more than once', id);
      this.loadingCount = 0;
    }
    delete this.state.messages[id];
    //* ********************* ???????????????????? immutable??????????
    this.setState({ shown: !!this.loadingCount, messages: this.state.messages });
  }
  clickHandler = () => {
    if (this.props.masked && this.props.closable) {
      this.loadingCount = 0;
      this.setState({ shown: false, messages: {} });
    }
  };

  render() {
    // console.log('this.props.theme', this.props.theme);
    const msgList = Object.keys(this.state.messages).length ? (
      <ul className={styles.messageList}>
        {Object.entries(this.state.messages).map(([k, v]) => <li key={k}>{v}</li>)}
      </ul>
    ) : (
      ''
    );
    // console.log('msgList', msgList);
    return (
      <div
        className={this.props.masked ? styles.globalMasked : styles.global}
        style={{ visibility: this.state.shown ? 'visible' : 'hidden' }}
        onClick={this.clickHandler}
      >
        <this.props.theme message={msgList} />
      </div>
    );
  }
}

export default class GlobalLoading {
  static singleInst = null;

  Comp = null;
  root = null;
  localHandled = false;
  constructor(theme, masked, closable) {
    if (this.constructor.singleInst) {
      // singleton: mount global once for ever; but still update its properties;
      this.render(theme, masked, closable);
      return this.constructor.singleInst;
    }
    this.constructor.singleInst = this;

    this.root = document.createElement('div');
    console.error('aaaaaaaaaaa', this.root);
    this.root.style.position = 'relative';
    this.render(theme, masked, closable);
    document.body.appendChild(this.root);
    LoadingEvent.on(EVENT_LOCAL_HADNDLED, this.flagLocalHandled);
  }
  flagLocalHandled = () => {
    this.localHandled = true;
  };
  render(theme, masked, closable) {
    aaaddafaaa;
    this.Comp = render(<GlobalLoadingComp {...{ theme, masked, closable }} />, this.root);
  }
  open(msg) {
    // wait for local loading to handle;
    const promise = new Promise((resolve, reject) => {
      window.setTimeout(() => {
        // console.log('this.localHandled', this.localHandled);
        if (this.localHandled) {
          // bypass this loading, and reset flag;
          this.localHandled = false;
          reject();
        } else {
          resolve(this.Comp.addLoading(msg));
        }
      }, 10);
    });
    promise.catch(() => {
      // console.log('global loading bypassed');
    }); // prevent error bubbling up;
    return promise;
  }
  close(promise) {
    if (!promise) return;
    promise.then(
      (id) => {
        this.Comp.removeLoading(id);
      },
      () => {
        // console.log('bypassed was not shown')
      },
    );
  }
}
