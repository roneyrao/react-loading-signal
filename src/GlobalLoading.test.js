import React from 'react';
import { shallow, mount } from 'enzyme';

import GlobalLoading, { GlobalLoadingComp } from './GlobalLoading';
import styles from './styles';
import * as Theme from './themes';

describe('GlobalLoading Component', () => {
  test('render properly with default props', () => {
    const wrapper = shallow(<GlobalLoadingComp />);
    expect(wrapper.prop('style').visibility).toBe('hidden');
    expect(wrapper.childAt(0).equals(<Theme.Spinning message='' />)).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });
  test('masked', () => {
    const wrapper = shallow(<GlobalLoadingComp masked />);
    expect(wrapper.hasClass(styles.globalMasked)).toBe(true);
  });
  test('closable', () => {
    const wrapper = shallow(<GlobalLoadingComp masked closable />);
    const inst = wrapper.instance();

    expect(wrapper.hasClass(styles.globalMasked)).toBe(true);
    expect(wrapper.prop('style').visibility).toBe('hidden');

    inst.addLoading('aaa');
    wrapper.update();
    expect(inst.loadingCount).toBe(1);
    expect(wrapper.prop('style').visibility).toBe('visible');

    wrapper.simulate('click');
    expect(inst.loadingCount).toBe(1);
    expect(Object.keys(inst.state.messages).length).toBe(1);
    expect(wrapper.prop('style').visibility).toBe('hidden');
  });
  test('custom theme', () => {
    const wrapper = shallow(<GlobalLoadingComp theme={Theme.Blobs} />);
    expect(wrapper.childAt(0).equals(<Theme.Blobs message='' />)).toBe(true);
  });
  describe('handle message processing', () => {
    const wrapper = mount(<GlobalLoadingComp />);
    const inst = wrapper.instance();
    const msg1 = 'this is a message';
    const msg2 = 'this is another message';

    let id1;
    let id2;

    test('addLoading', () => {
      id1 = inst.addLoading(msg1);
      wrapper.update();

      expect(inst.loadingCount).toBe(1);
      expect(inst.state.shown).toBe(true);
      expect(Object.keys(inst.state.messages).length).toBe(1);
      expect(wrapper.getDOMNode().style.visibility).toBe('visible');
      const ul = wrapper.find('ul');
      expect(ul.exists()).toBe(true);
      expect(ul.children().length).toBe(1);
      expect(ul.childAt(0).text()).toBe(msg1);

      id2 = inst.addLoading(msg2);
      wrapper.update();
      expect(wrapper
        .find('ul')
        .childAt(1)
        .text()).toBe(msg2);
    });

    test('removeLoading', () => {
      inst.removeLoading(id1);
      wrapper.update();
      expect(wrapper
        .find('ul')
        .childAt(0)
        .text()).toBe(msg2);

      inst.removeLoading(id2);
      wrapper.update();
      expect(wrapper.find('ul').exists()).toBe(false);
      expect(wrapper.getDOMNode().style.visibility).toBe('hidden');
      expect(inst.loadingCount).toBe(0);

      inst.removeLoading(id2);
      expect(inst.loadingCount).toBe(0);
    });
  });
});

describe('GlobalLoading Class', () => {
  let inst;
  const id = 'msg_id';
  let addLoading;
  let removeLoading;
  beforeAll(() => {
    inst = new GlobalLoading(undefined, true);
    inst.Comp.addLoading = jest.fn(() => id);
    ({ addLoading } = inst.Comp);
    inst.Comp.removeLoading = jest.fn();
    ({ removeLoading } = inst.Comp);
  });

  test('component is rendered into body', () => {
    expect(document.body.querySelector(`.${styles.loading}`)).not.toBeNull();
  });
  test('component is singleton', () => {
    const first = inst.constructor.singleInst;
    const inst2 = new GlobalLoading(null, true);
    expect(first).toBe(inst2);
  });
  test('get singleton', () => {
    expect(GlobalLoading.getInstance()).toBe(inst);
  });
  test('add and remove message', (done) => {
    const msg = 'loading message';
    const prom = inst.open(msg);
    setTimeout(() => {
      try {
        expect(addLoading).toBeCalledWith(msg);

        inst.close(prom);
        prom
          .then(
            () => {
              expect(removeLoading).toBeCalledWith(id);
              done();
            },
            (err) => {
              done.fail(err);
            },
          )
          .catch(err => done.fail(err));
      } catch (err) {
        done.fail(err);
      }
    }, 100);
  });
  test('bypass a message', (done) => {
    addLoading.mockReset();
    removeLoading.mockReset();

    const msg = 'loading message';
    const prom = inst.open(msg);
    prom._canceled = true;

    setTimeout(() => {
      try {
        expect(addLoading).not.toBeCalled();

        inst.close(prom);
        prom
          .then(
            () => {
              done.faile('never call this');
            },
            () => {
              expect(removeLoading).not.toBeCalled();
              done();
            },
          )
          .catch(err => done.fail(err));
      } catch (err) {
        done.fail(err);
      }
    }, 100);
  });
});
