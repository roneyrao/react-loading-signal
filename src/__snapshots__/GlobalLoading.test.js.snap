// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalLoading Component render properly with default props 1`] = `
<div
  className="FlexLoading__global FlexLoading__loading"
  onClick={[Function]}
  style={
    Object {
      "visibility": "hidden",
    }
  }
>
  <LoadingSpinning
    message=""
  />
</div>
`;
